Mathlib/LinearAlgebra/Reflection.lean

```diff
@@ -0,0 +1,161 @@
+/-
+Copyright (c) 2023 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash, Deepro Choudhury
+-/
+import Mathlib.GroupTheory.OrderOfElement
+import Mathlib.LinearAlgebra.Dual
+import Mathlib.LinearAlgebra.FiniteSpan
+
+/-!
+# Reflections in linear algebra
+
+Given an element `x` in a module `M` together with a linear form `f` on `M` such that `f x = 2`, the
+map `y ↦ y - (f y) • x` is an involutive endomorphism of `M`, such that:
+ 1. the kernel of `f` is fixed,
+ 2. the point `x ↦ -x`.
+
+Such endomorphisms are often called reflections of the module `M`. When `M` carries a inner product
+for which `x` is perpendicular to the kernel of `f`, then (with mild assumptions) the endomorphism
+is characterised by properties 1 and 2 above, and is a linear isometry.
+
+## Main definitions / results:
+
+ * `Module.preReflection`: the definition of the map `y ↦ y - (f y) • x`. Its main utility lies in
+   the fact that it does not require the assumption `f x = 2`, giving the user freedom to defer
+   discharging this proof obligation.
+ * `Module.reflection`: the definition of the map `y ↦ y - (f y) • x`. This requires the assumption
+   that `f x = 2` but by way of compensation it produces a linear equivalence rather than a mere
+   linear map.
+ * `Module.Dual.eq_of_preReflection_image_subset`: a uniqueness result about reflections preserving
+   finite spanning sets that is useful in the theory of root data / systems.
+
+## TODO
+
+Related definitions of reflection exists elsewhere in the library. These more specialised
+definitions, which require an ambient `InnerProductSpace` structure, are `reflection` (of type
+`LinearIsometryEquiv`) and `EuclideanGeometry.reflection` (of type `AffineIsometryEquiv`). We
+should connect (or unify) these definitions with `Module.reflecton` defined here.
+
+-/
+
+open Set Function Pointwise
+open Module hiding Finite
+open Submodule (span)
+
+noncomputable section
+
+variable {R M : Type*} [CommRing R] [AddCommGroup M] [Module R M] (x y : M) (f : Dual R M)
+
+namespace Module
+
+/-- Given an element `x` in a module `M` and a linear form `f` on `M`, we define the endomorphism
+of `M` for which `y ↦ y - (f y) • x`.
+
+One is typically interested in this endomorphism when `f x = 2`; this definition exists to allow the
+user defer discharging this proof obligation. See also `Module.reflection`. -/
+def preReflection : End R M :=
+  LinearMap.id - f.smulRight x
+
+lemma preReflection_apply :
```

Comment:
Is it a conscious design choice that the order of arguments for this lemma will be `x y f`? To me it feels like `x` and `f` belong together, more so than `x` and `y`.

Reply:
Agreed, I have changed the order.